File: 백준/Silver/1931. 회의실 배정/회의실 배정.cc
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

vector<pair<int, int>> arr;
vector<pair<int, int>> ans;

bool comp(const pair<int, int>& a, const pair<int, int>& b) {
  if(a.second == b.second) return a.first<b.first;
  return a.second < b.second;
}
int main(){
  int n;
  cin>>n;
  for(int i=0;i<n;i++){
    int a,b;
    cin>>a>>b;
    arr.push_back(make_pair(a,b));
  }

  sort(arr.begin(), arr.end(), comp);

  ans.push_back(make_pair(arr.front().first, arr.front().second));

  for(int i=1;i<arr.size();i++){
    if(ans[ans.size()-1].second<=arr[i].first){
      ans.push_back(make_pair(arr[i].first, arr[i].second));
    }
  }

  // for(int i=0;i<ans.size();i++){
  //   cout<<ans[i].first<<" "<<ans[i].second<<endl;
  // }
  cout<<ans.size();

  
}

File: 백준/Silver/1931. 회의실 배정/회의실 배정.java
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    public static void main(String[] args) throws Exception {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int num = Integer.parseInt(br.readLine());
        List<Pair> list = new ArrayList<>();         
        for (int i = 0; i < num; i++) {             
            String[] tmp = br.readLine().split(" ");             
            int first = Integer.parseInt(tmp[0]);             
            int second = Integer.parseInt(tmp[1]);             
            list.add(new Pair(first, second));         
        }
        Collections.sort(list, new Comparator<Pair>() {             
            @Override             
            public int compare(Pair x, Pair y) {
                if(x.y==y.y) return x.x-y.x;
                return x.y - y.y;
            }         
        });
        // for(Pair p: list){
        //     System.out.println(p.x+" "+p.y);
        // }
        
        int end = list.get(0).y;
        int cnt = 1;
        
        for(int i=1;i<num;i++){
           if(end <= list.get(i).x){
               end = list.get(i).y;
               cnt++;
           }
        }
        
        System.out.println(cnt);
        
    }
    
    public static class Pair{
        int x;
        int y;
        Pair(int x, int y){
            this.x = x;
            this.y = y;
        }
    }
    
}
